File: Ghostlark/Sources/Core/ClashAPI.swift
import Foundation

/// Minimal client for sing-box's Clash-compatible control API.
struct ClashAPI {
    let port: Int
    let secret: String

    struct APIError: LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    /// A session that ignores the system proxy so control traffic never loops through the tunnel.
    static let session: URLSession = {
        let c = URLSessionConfiguration.ephemeral
        c.connectionProxyDictionary = [:]
        c.timeoutIntervalForRequest = 30
        c.httpMaximumConnectionsPerHost = 64
        return URLSession(configuration: c)
    }()

    private func request(_ path: String, timeout: TimeInterval = 30) -> URLRequest {
        var r = URLRequest(url: URL(string: "http://127.0.0.1:\(port)\(path)")!)
        r.setValue("Bearer \(secret)", forHTTPHeaderField: "Authorization")
        r.timeoutInterval = timeout
        return r
    }

    func waitReady(timeout: TimeInterval) async -> Bool {
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            if let (_, resp) = try? await ClashAPI.session.data(for: request("/version", timeout: 2)),
               (resp as? HTTPURLResponse)?.statusCode == 200 { return true }
            try? await Task.sleep(nanoseconds: 150_000_000)
        }
        return false
    }

    /// Runs a connectivity test through one outbound. Returns latency in ms.
    func delay(tag: String, url: String, timeoutMs: Int) async throws -> Int {
        var comps = URLComponents(string: "http://127.0.0.1:\(port)/proxies/\(tag)/delay")!
        comps.queryItems = [URLQueryItem(name: "timeout", value: String(timeoutMs)), URLQueryItem(name: "url", value: url)]
        var r = URLRequest(url: comps.url!)
        r.setValue("Bearer \(secret)", forHTTPHeaderField: "Authorization")
        r.timeoutInterval = TimeInterval(timeoutMs) / 1000 + 5
        let (data, resp) = try await ClashAPI.session.data(for: r)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        guard code == 200, let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let d = obj["delay"] as? Int else {
            let msg = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["message"] as? String
            throw APIError(message: code == 504 ? "timeout" : (msg ?? "HTTP \(code)"))
        }
        return d
    }

    /// The member a urltest/selector group is currently using.
    func groupNow(tag: String) async -> String? {
        guard let (data, _) = try? await ClashAPI.session.data(for: request("/proxies/\(tag)", timeout: 5)),
              let o = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        return o["now"] as? String
    }

    /// Streams (upBytesPerSec, downBytesPerSec) samples.
    func trafficStream() -> AsyncThrowingStream<(Int, Int), Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    let (bytes, _) = try await ClashAPI.session.bytes(for: request("/traffic", timeout: 3600 * 24))
                    for try await line in bytes.lines {
                        if let d = line.data(using: .utf8),
                           let o = try? JSONSerialization.jsonObject(with: d) as? [String: Any] {
                            continuation.yield((o["up"] as? Int ?? 0, o["down"] as? Int ?? 0))
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}

File: Android_Version/app/src/main/java/com/sepahi/ghostlark/core/ClashApi.kt
package com.sepahi.ghostlark.core

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.Proxy
import java.net.URL
import java.net.URLEncoder

/** Minimal client for sing-box's Clash-compatible control API on loopback. */
class ClashApi(private val port: Int, private val secret: String) {

    private fun open(path: String, timeoutMs: Int): HttpURLConnection {
        val c = URL("http://127.0.0.1:$port$path").openConnection(Proxy.NO_PROXY) as HttpURLConnection
        c.setRequestProperty("Authorization", "Bearer $secret")
        c.connectTimeout = 2000
        c.readTimeout = timeoutMs
        return c
    }

    suspend fun waitReady(timeoutMs: Long): Boolean = withContext(Dispatchers.IO) {
        val deadline = System.currentTimeMillis() + timeoutMs
        while (System.currentTimeMillis() < deadline) {
            try {
                val c = open("/version", 2000)
                if (c.responseCode == 200) { c.disconnect(); return@withContext true }
                c.disconnect()
            } catch (_: Exception) { }
            delay(150)
        }
        false
    }

    /** Connectivity test through one outbound; returns latency in ms or throws. */
    suspend fun delay(tag: String, url: String, timeoutMs: Int): Int = withContext(Dispatchers.IO) {
        val c = open("/proxies/$tag/delay?timeout=$timeoutMs&url=${URLEncoder.encode(url, "UTF-8")}", timeoutMs + 5000)
        try {
            val code = c.responseCode
            val body = (if (code == 200) c.inputStream else c.errorStream)?.bufferedReader()?.readText() ?: ""
            if (code == 200) JSONObject(body).getInt("delay")
            else throw IOException(if (code == 504) "timeout" else runCatching { JSONObject(body).getString("message") }.getOrDefault("HTTP $code"))
        } finally { c.disconnect() }
    }

    /** Streams (upBytesPerSec, downBytesPerSec). */
    fun traffic(): Flow<Pair<Long, Long>> = flow {
        val c = open("/traffic", 0)
        try {
            c.inputStream.bufferedReader().useLines { lines ->
                for (line in lines) {
                    val o = runCatching { JSONObject(line) }.getOrNull() ?: continue
                    emit(o.optLong("up") to o.optLong("down"))
                }
            }
        } finally { c.disconnect() }
    }.flowOn(Dispatchers.IO)
}
